src/main/java/org/opensearch/knn/indices/ModelDao.java

```diff
@@ -427,61 +443,187 @@ private String getMapping() throws IOException {
+                OpenSearchException exception = null;
+                if (removeModelFromCacheResponse.hasFailures()) {
+                    String failureMessage = buildRemoveModelErrorMessage(modelId, removeModelFromCacheResponse);
+                    exception = new OpenSearchException(failureMessage);
+                }
 
-                        listener.onResponse(new DeleteModelResponse(modelId, "failed", failureMessage));
+                // Remove modelId from model graveyard
+                updateModelGraveyardToDelete(modelId, true, unblockModelIdStep, exception);
 
-                    }, e -> listener.onResponse(new DeleteModelResponse(modelId, "failed", e.getMessage())))
-                );
-            }, e -> listener.onResponse(new DeleteModelResponse(modelId, "failed", e.getMessage())));
+            }, e -> listener.onFailure(new OpenSearchException(e)));
 
-            // On model metadata removal, delete the model from the index
-            ActionListener<AcknowledgedResponse> onMetadataUpdateListener = ActionListener.wrap(
-                acknowledgedResponse -> deleteRequestBuilder.execute(onModelDeleteListener),
-                listener::onFailure
+            unblockModelIdStep.whenComplete(acknowledgedResponse -> {
+                // After clearing the cache, if there are no errors return the response
+                listener.onResponse(new DeleteModelResponse(modelId, DELETED, null));
+
+            }, listener::onFailure);
+
+        }
+
+        // Remove model from cache in the cluster
+        private void removeModelFromCache(String modelId, StepListener<RemoveModelFromCacheResponse> clearModelFromCacheStep) {
+            client.execute(
+                RemoveModelFromCacheAction.INSTANCE,
+                new RemoveModelFromCacheRequest(modelId),
+                ActionListener.wrap(
+                    clearModelFromCacheStep::onResponse,
+                    exception -> removeModelIdFromGraveyardOnFailure(modelId, exception, clearModelFromCacheStep)
+                )
             );
+        }
+
+        // Delete model from the system index
+        private void deleteModelFromIndex(
+            String modelId,
+            StepListener<DeleteResponse> deleteModelFromIndexStep,
+            DeleteRequestBuilder deleteRequestBuilder
+        ) {
+            deleteRequestBuilder.execute(
+                ActionListener.wrap(
+                    deleteModelFromIndexStep::onResponse,
+                    exception -> removeModelIdFromGraveyardOnFailure(modelId, exception, deleteModelFromIndexStep)
+                )
+            );
+        }
+
+        // Update model graveyard to add/remove modelId
+        private void updateModelGraveyardToDelete(
+            String modelId,
+            boolean isRemoveRequest,
+            StepListener<AcknowledgedResponse> step,
+            Exception exception
```

Comment:
Can we use Optional<Exception> here? Seems that there are many cases possible when we're passing null, I think Optional fits such use case better.